sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,348 @@
+                self.data.as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry,
+                len,
+            )
+        }
+    }
+
+    /// Returns a reference to the entry at `index` **without** bounds checking.
+    ///
+    /// # Safety
+    /// Caller must guarantee that `index < self.len()`.
+    #[inline(always)]
+    pub unsafe fn get_entry_unchecked(&self, index: usize) -> &SlotHashEntry {
+        debug_assert!(index < self.len());
+        &self.as_entries_slice()[index]
+    }
+}
+
+impl<'a, T: Deref<Target = [u8]>> IntoIterator for &'a SlotHashes<T> {
+    type Item = &'a SlotHashEntry;
+    type IntoIter = core::slice::Iter<'a, SlotHashEntry>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.as_entries_slice().iter()
+    }
+}
+
+impl<'a> SlotHashes<Ref<'a, [u8]>> {
+    /// Creates a `SlotHashes` instance by safely borrowing data from an `AccountInfo`.
+    ///
+    /// This function verifies that:
+    /// - The account key matches the `SLOTHASHES_ID`
+    #[inline(always)]
+    pub fn from_account_info(account_info: &'a AccountInfo) -> Result<Self, ProgramError> {
+        if account_info.key() != &SLOTHASHES_ID {
+            return Err(ProgramError::Custom(ERR_WRONG_ACCOUNT_KEY));
+        }
+
+        let data_ref = account_info.try_borrow_data()?;
+
+        // SAFETY: The account was validated to be the `SlotHashes` sysvar.
+        Ok(unsafe { SlotHashes::new_unchecked(data_ref) })
+    }
+}
+
+#[cfg(feature = "std")]
+impl SlotHashes<Box<[u8]>> {
+    /// Allocates a buffer and fetches SlotHashes sysvar data via syscall.
+    ///
+    /// # Safety
+    /// The caller must ensure the buffer pointer is valid for MAX_SIZE bytes.
+    /// The syscall will write exactly MAX_SIZE bytes to the buffer.
+    #[inline(always)]
+    unsafe fn fetch_into_buffer(buffer_ptr: *mut u8) -> Result<(), ProgramError> {
+        crate::sysvars::get_sysvar_unchecked(buffer_ptr, &SLOTHASHES_ID, 0, MAX_SIZE)?;
+
+        // For tests on builds that don't actually fill the buffer.
+        #[cfg(not(target_os = "solana"))]
+        core::ptr::write_bytes(buffer_ptr, 0, NUM_ENTRIES_SIZE);
+
+        Ok(())
+    }
+
+    /// Allocates an optimal buffer for the sysvar data based on available features.
+    #[inline(always)]
+    fn allocate_and_fetch() -> Result<Box<[u8]>, ProgramError> {
+        // Prefer the zero-init-skip API when available (Rust ≥1.82) but
+        // transparently fall back to a `Vec` for older compilers.
+
+        #[cfg(has_box_new_uninit_slice)]
+        #[allow(clippy::incompatible_msrv)]
+        {
+            // SAFETY: The buffer length matches the requested syscall length and we
+            // fully initialise it before use.
+            let mut data = Box::new_uninit_slice(MAX_SIZE);
+            unsafe {
+                Self::fetch_into_buffer(data.as_mut_ptr() as *mut u8)?;
+                Ok(data.assume_init())
+            }
+        }
+
+        #[cfg(not(has_box_new_uninit_slice))]
```

Comment:
This probably can go once the rust version is bumped, together with lines 307 and 308.